File: java/22._GenerateParentheses.java
public class Solution {
    public static List<String> generateParenthesis(int n) {
    	List<String> res = new ArrayList<String>();
    	handdle(n, n, "", res);
    	return res;
    }
    public static void handdle(int l, int r, String s, List<String> res) {
    	if(l < 0 || r < 0 || l > r) {
    		return;
    	}
    	if(l == 0 && r == 0) {
    		res.add(s);
    	}
    	handdle(l - 1, r, s + "(", res);
    	handdle(l, r - 1, s + ")", res);
    }
}

File: c++/22_GenerateParentheses.cpp
class Solution {
public:
    vector<string> res;
    vector<string> generateParenthesis(int n) {
        dfs(n, 0, 0, "");
        return res;
    }

    void dfs(int n, int lc, int rc, string path) {
        if (lc == n && rc == n) {
            res.push_back(path);
            return;
        }
        if (lc < n) dfs(n, lc + 1, rc, path + '(');
        if (rc < n && lc > rc) dfs(n, lc, rc + 1, path + ')');
    }
};